fms_extras/utils/generation.py

```diff
@@ -0,0 +1,155 @@
+import functools
+import time
+from typing import Any, Callable, List, MutableMapping, Optional, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+
+from fms_extras.utils.cache.paged import PagedKVCacheManager
+
+
+def __right_pad_zeros(input_tensor: torch.Tensor, max_length: int) -> torch.Tensor:
+    return torch.stack(
+        [
+            F.pad(
+                input_tensor[i],
+                (0, max_length - input_tensor.size(1)),
+            )
+            for i in range(input_tensor.size(0))
+        ]
+    )
+
+
+def paged_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids_list: List[torch.Tensor],
+    kv_cache_manager: PagedKVCacheManager,
+    max_seq_len: int = 2048,
+    max_new_tokens: int = 256,
+    temperature: float = 1.0,
+    top_k: int = 10,
+    do_sample: bool = True,
+    decode_model: Optional[Union[Callable, torch.nn.Module]] = None,
+    # todo: This is a WIP to enable cudagraphs, currently its only for batch_size=1
+    cudagraphs: bool = False,
+) -> Tuple[torch.Tensor, int, float, float]:
+    """
+    A trivial generate function that can be used for validation/testing generation using paged attention
+
+    Args:
+        model: Callable or nn.Module
+            A function or nn.Module that takes a batch of input_ids and returns logits
+        input_ids_list: List[torch.Tensor]
+            A list of tensors, each tensor being for a single prompt
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+        cudagraphs: bool
+            if True, model is using cudagraphs and input will be padded, otherwise no cudagraphs is used and input is
+            not padded
+
+    Returns:
+    Tuple[torch.Tensor, int, float, float]
+        the resulting output tokens, the number of new tokens generated, the time to first token in seconds and the
+        total decode time
```

Comment:
I'm not sure "decode time" clearly indicates that we mean total time to generate subsequent tokens after kv-cache is populated

Reply:
Fixed in latest commit